ta/pkcs11/src/pkcs11_token.c

```diff
@@ -80,3 +81,96 @@ void pkcs11_deinit(void)
 	for (id = 0; id < TOKEN_COUNT; id++)
 		close_persistent_db(get_token(id));
 }
+
+uint32_t entry_ck_slot_list(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *out = &params[2];
+	const size_t out_size = sizeof(uint32_t) * TOKEN_COUNT;
+	uint8_t *id = NULL;
+	uint32_t n = 0;
+
+	if (ptypes != exp_pt ||
+	    params[0].memref.size != TEE_PARAM0_SIZE_MIN)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	if (out->memref.size < out_size) {
+		out->memref.size = out_size;
+
+		if (out->memref.buffer)
+			return PKCS11_CKR_BUFFER_TOO_SMALL;
+		else
+			return PKCS11_CKR_OK;
+	}
+
+	for (id = out->memref.buffer, n = 0; n < TOKEN_COUNT;
+	     n++, id += sizeof(n))
+		TEE_MemMove(id, &n, sizeof(n));
+
+	out->memref.size = out_size;
+
+	return PKCS11_CKR_OK;
+}
+
+uint32_t entry_ck_slot_info(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = &params[0];
+	TEE_Param *out = &params[2];
+	uint32_t rv = 0;
+	struct serialargs ctrlargs = { };
+	uint32_t token_id = 0;
+	struct ck_token *token = NULL;
+	struct pkcs11_slot_info info = {
+		.slot_description = PKCS11_SLOT_DESCRIPTION,
+		.manufacturer_id = PKCS11_SLOT_MANUFACTURER,
+		.flags = PKCS11_CKFS_TOKEN_PRESENT,
+		.hardware_version = PKCS11_SLOT_HW_VERSION,
+		.firmware_version = PKCS11_SLOT_FW_VERSION,
+	};
+	int n = 0;
+
+	COMPILE_TIME_ASSERT(sizeof(PKCS11_SLOT_DESCRIPTION) <=
+			    sizeof(info.slot_description));
+	COMPILE_TIME_ASSERT(sizeof(PKCS11_SLOT_MANUFACTURER) <=
+			    sizeof(info.manufacturer_id));
+
+	if (ptypes != exp_pt ||
+	    out->memref.size != sizeof(struct pkcs11_slot_info))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rv = serialargs_get(&ctrlargs, &token_id, sizeof(uint32_t));
```

Comment:
For consistency, maybe prefer s/`sizeof(uint32_t)`/`sizeof(token_id)`/

And equivalent changes at lines 145 and 172.

Reply:
Yes, I think that's better.